### PR TITLE
Coupon Management: List - CouponListViewController

### PR DESCRIPTION
**Pull request by joshheald-a8c**
Part of #3916 

## Description
Adds a basic list of coupons to the coupon list view controller:
 - loading state
 - load and display the first page of coupons (only) 
 - an unfinished empty state (requires correct copy, image, and button implementation).

This is clearly a work in progress, given the above description!

## Testing
Open Settings, then tap "Manage Coupons" in the list.
Observe the loading state, and then that coupon cells replace the loading state if any are present on the store.
If there are no coupons, an (unfinished) empty state is displayed.
If there is an error, this is not yet displayed to the user, and an empty list is shown.

### Coupon list

https://user-images.githubusercontent.com/80399824/116983320-d8268200-acc1-11eb-8740-bb27e991ccba.mp4

### Empty state

https://user-images.githubusercontent.com/80399824/116983561-2cc9fd00-acc2-11eb-86f5-6f30a43f8893.mp4

Update release notes:
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

**Comment by joshheald-a8c** [comment]
@pmusolino Sorry this one has gotten a little long, since it's over 500 lines. I considered splitting out the empty/error state stuff to reduce it, but in the end I thought it was better to include them together so you can see the approach now and let me know if you think I should make changes to it.

In particular, I'm starting to think about the paging and whether my original plan was flawed by not using `PaginatedListViewControllerStateCoordinator` either directly on the view controller, or as a part of the view model's interface. The `didLeave` and `didEnter` approach it has perhaps gives a bit more context to correctly update the view, compared to the more implicit interpretation the view has to make of the single piece of state change information that the coupon list view model currently exposes to it. I.e. currently, the clearing down of placeholder cells and the empty state has to happen at different places in `viewModelDidChange`.

Would love to get your thoughts on this in particular.

File: WooCommerce/Classes/Extensions/UIImage+Woo.swift
Comment (ealeksandrov) [inline WooCommerce/Classes/Extensions/UIImage+Woo.swift]: For all images here we also add a test case to `IconsTests` to make sure all assets included.
And since it's a temp placehloder - maybe just reuse `errorImage` directly for now?

File: WooCommerce/Classes/ViewRelated/Coupons/CouponManagementListViewModel.swift
Comment (ealeksandrov) [inline WooCommerce/Classes/ViewRelated/Coupons/CouponManagementListViewModel.swift]: Why do we need `initialized` state? Can we merge it with empty?

File: WooCommerce/Classes/ViewRelated/Coupons/CouponManagementListViewModel.swift
Comment (ealeksandrov) [inline WooCommerce/Classes/ViewRelated/Coupons/CouponManagementListViewModel.swift]: nit: `isEmpty == false` would fit better here

File: WooCommerce/Classes/ViewRelated/Coupons/CouponManagementViewController.swift
Comment (ealeksandrov) [inline WooCommerce/Classes/ViewRelated/Coupons/CouponManagementViewController.swift]: Can we make it `private`?

File: WooCommerce/Classes/ViewRelated/Coupons/CouponManagementViewController.swift
Comment (ealeksandrov) [inline WooCommerce/Classes/ViewRelated/Coupons/CouponManagementViewController.swift]: nit: no need for `self`

**Review by ealeksandrov — CHANGES_REQUESTED** [review]
Hey @joshheald-a8c!

I've tested and confirm that I can see existing coupons in a list.
But when I don't have any coupons - I don't see empty state, it just loads endlessly (while I get `Synchronized coupons` in a console).

https://user-images.githubusercontent.com/3132438/117136121-9b858400-adb0-11eb-8c93-4669e85d5ce8.mp4

> whether my original plan was flawed by not using PaginatedListViewControllerStateCoordinator

`PaginatedListViewControllerStateCoordinator` may be a good idea, since it provides a dedicated callback to manage state of a view - I think I would choose it. Of course it's possible to update state handling in your viewmodel with pagination integration in mind.

> either directly on the view controller, or as a part of the view model's interface

I would also keep resultsController in a VC since this screen doesn't have much logic to encapsulate it in a VM (since we already have dedicated solutions for data fetching). There is just view manipulation in response to different contents types.

Also one thing that caught my attention - there are a few autolayout errors when I open coupons lists (`Unable to simultaneously satisfy constraints.`)

File: WooCommerce/Classes/Extensions/UIImage+Woo.swift
Comment (joshheald-a8c) [reply WooCommerce/Classes/Extensions/UIImage+Woo.swift]: Thanks for the heads up. I have an image now but to avoid this PR getting bigger I'll switch to errorImage for the time being and update the empty state in a future PR, with tests for the new image.

File: WooCommerce/Classes/ViewRelated/Coupons/CouponManagementListViewModel.swift
Comment (joshheald-a8c) [reply WooCommerce/Classes/ViewRelated/Coupons/CouponManagementListViewModel.swift]: If we want to go get rid of it, it might be better to merge it with loading, because briefly showing an empty state first could be misleading...

`initialized` was added so that we could represent all the possible states of the screen: when we first init, we're not yet loading, we don't know whether there are coupons or not, etc. It's a habit from the RxSwift architecture we used for a while at my current job, definitely not something we need to keep here.

It's not used like this in this example, but it can be helpful to inform the view controller that the view model is ready for use in some situations. I'll remove it for now and we can add it back if it turns out it's needed.

**Comment by joshheald-a8c** [comment]
Thanks for your review @ealeksandrov 👍 

I've fixed up the nits, smaller changes, and runtime constraint warnings just now. Further changes still to make on the other stuff and I'll let you know when I'm done with those...

> But when I don't have any coupons - I don't see empty state, it just loads endlessly (while I get `Synchronized coupons` in a console).
> `PaginatedListViewControllerStateCoordinator` may be a good idea, since it provides a dedicated callback to manage state of a view - I think I would choose it. Of course it's possible to update state handling in your viewmodel with pagination integration in mind.

I think so too, especially given the issue you've found with the empty state. I'll switch to using the PaginatedListCoordinator with another commit later on, and resolve that empty state bug at the same time.

> I would also keep resultsController in a VC since this screen doesn't have much logic to encapsulate it in a VM (since we already have dedicated solutions for data fetching). There is just view manipulation in response to different contents types.

This one I'm less sure on, and I'm going to think more about it... if it's in the VC, it'll be harder to test the cell view model creation, which will have more logic around it as I move through the work here.

**Comment by joshheald-a8c** [comment]
@ealeksandrov when you have time, I'd really appreciate a follow up review on this please :)

> I'll switch to using the PaginatedListCoordinator with another commit later on, and resolve that empty state bug at the same time.
So, I tried this, and ended up deciding against using it directly. I found that when I moved to try implementing pull to refresh, and loading subsequent pages, using the existing coordinator meant for either: 
1. complicated state translation code in the view model, or
2. complicated untestable state handling code in the view controller

This was due to `.syncing` representing three distinct states: first page load, subsequent page loads, and refreshing after loading data.

Instead, I've opted to take inspiration from the implementation of the existing coordinator, with the intention to extend it to cover these three states explicitly in a later commit. I think that should be done as a refactor after the coupon work is merged, because it will touch unrelated screens and we'd run the risk of causing either merge conflicts, or accidentally breaking code which is added after the coupon branch was taken.

This has also meant a small change to my initial plan, where I now have two closures on the view model (didEnterState, didLeaveState) to update the state of the view rather than one (onStateChange). This means we no longer have to rely on interpreting one state to mean that we've left another, and really cleans up the view controller.

What do you think of these changes?

File: WooCommerce/Classes/ViewRelated/Coupons/CouponManagementListViewModel.swift
Comment (joshheald-a8c) [reply WooCommerce/Classes/ViewRelated/Coupons/CouponManagementListViewModel.swift]: I ended up bringing this back. Having an explicit change from `.initialized` to `.loading` gives the view controller an explicit `didEnter` state change after it tells the view model to start loading. This gives a straightforward way to start showing the placeholder cells, without the risk of showing the empty state briefly.

File: WooCommerce/Classes/ViewRelated/Coupons/CouponManagementListViewModel.swift
Comment (ealeksandrov) [inline WooCommerce/Classes/ViewRelated/Coupons/CouponManagementListViewModel.swift]: nit:
```suggestion
    var accessibilityLabel: String
```

File: WooCommerce/Classes/ViewRelated/Coupons/CouponManagementListViewModel.swift
Comment (ealeksandrov) [inline WooCommerce/Classes/ViewRelated/Coupons/CouponManagementListViewModel.swift]: nit:
```suggestion
    /// The ViewController calls `viewDidLoad` to notify the view model it's ready to receive results
```

File: WooCommerce/Classes/ViewRelated/Coupons/CouponManagementListViewModel.swift
Comment (ealeksandrov) [inline WooCommerce/Classes/ViewRelated/Coupons/CouponManagementListViewModel.swift:176]: nit: `success` case should come first, before failure

File: WooCommerce/Classes/ViewRelated/Dashboard/Settings/SettingsViewController.swift
Comment (ealeksandrov) [inline WooCommerce/Classes/ViewRelated/Dashboard/Settings/SettingsViewController.swift]: Since it's used only in a single place, what do you think about moving it there? Also fallback to 0 will break things on CouponManagementViewController, maybe add guard with error message instead?

File: WooCommerce/Classes/ViewRelated/Coupons/CouponManagementViewController.swift
Comment (ealeksandrov) [inline WooCommerce/Classes/ViewRelated/Coupons/CouponManagementViewController.swift]: nit: there is `pinSubviewToAllEdges` helper that can simplify this

File: WooCommerce/Classes/ViewRelated/Coupons/CouponManagementViewController.swift
Comment (ealeksandrov) [inline WooCommerce/Classes/ViewRelated/Coupons/CouponManagementViewController.swift]: very nit: here I would prefer 2 braces without newline between. And there are more cases of similar spaced closing braces above.

I've spend some time to find a rule about it and I think it aligns with https://github.com/raywenderlich/swift-style-guide/#spacing:

> There should be no blank lines after an opening brace or before a closing brace.

(don't assume strict style requirements, I've just noticed inconsistency between VC and VM looking side-by-side 🙂)

File: WooCommerce/Classes/ViewRelated/Coupons/CouponManagementViewController.swift
Comment (ealeksandrov) [inline WooCommerce/Classes/ViewRelated/Coupons/CouponManagementViewController.swift]: Any reason for `IBOutlet` to be optional?

**Review by ealeksandrov — APPROVED** [review]
Nice work @joshheald-a8c!
I can see empty, loading and content state fine ✅
State logic works well and looks good ✅

I've added more smaller nits, since I've looked more line-by-line this time, but there is nothing critical.
To catch more typos - check `Edit > Format > Spelling and Grammar` in Xcode, I've learned about it recently and it's awesome 😄

Ping me after you push last bits (if you wish) and ready to merge it.

Following our `initialized` case discussion - I've tried to replace it with both `empty` and `coupons` and it worked well for me, no glitches or anything. And there is no logic around it, it's just initial placeholder value 🤔
I'll let you decide on keeping it in next iteration of this, maybe you'll find it useful in a more complete flow.

File: WooCommerce/Classes/ViewRelated/Coupons/CouponManagementListViewModel.swift
Comment (joshheald-a8c) [reply WooCommerce/Classes/ViewRelated/Coupons/CouponManagementListViewModel.swift]: Thanks, good spot!

File: WooCommerce/Classes/ViewRelated/Coupons/CouponManagementListViewModel.swift
Comment (joshheald-a8c) [reply WooCommerce/Classes/ViewRelated/Coupons/CouponManagementListViewModel.swift]: And again 🦅 👁️

File: WooCommerce/Classes/ViewRelated/Dashboard/Settings/SettingsViewController.swift
Comment (joshheald-a8c) [reply WooCommerce/Classes/ViewRelated/Dashboard/Settings/SettingsViewController.swift]: I've added a pop up error message if there is no store ID, with an action to let the user select the store. Was this what you meant for the error message? I think it's probably better than a `fatalError` in this instance, since it should be user-recoverable really.
coupon-list-opening-error

File: WooCommerce/Classes/ViewRelated/Coupons/CouponManagementViewController.swift
Comment (joshheald-a8c) [reply WooCommerce/Classes/ViewRelated/Coupons/CouponManagementViewController.swift]: Thanks!

File: WooCommerce/Classes/ViewRelated/Coupons/CouponManagementViewController.swift
Comment (joshheald-a8c) [reply WooCommerce/Classes/ViewRelated/Coupons/CouponManagementViewController.swift]: Yeah, the inconsistency wasn't great. I've removed the excess newlines and prefer it without them too. 

I think I've taken different cues from different existing files as I've been working, but hadn't spotted that in the RW style guide. Thanks for finding that 🙇 it's a well stated guideline, nice and clear.

File: WooCommerce/Classes/ViewRelated/Coupons/CouponManagementViewController.swift
Comment (joshheald-a8c) [reply WooCommerce/Classes/ViewRelated/Coupons/CouponManagementViewController.swift]: Nope, just something which was useful earlier on in writing it but not needed any more. Removed 😃

**Comment by joshheald-a8c** [comment]
@ealeksandrov Thanks for your review 👍 
I've fixed up the remaining nits and improvements you identified, so whenever you're ready to take a look, that would be awesome 😁

**Review by ealeksandrov — APPROVED** [review]
LGTM :shipit:

BTW, that's a nice implementation of a error for "please select a store" - I was thinking about simple log into console 😅

**Pull request merged**
